optax/contrib/_sophia_h.py

```diff
@@ -0,0 +1,307 @@
+# Copyright 2023 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Sophia-H optimizer.
+
+A contributed implementation of the Sophia-H optimizer from "Sophia: A Scalable
+Stochastic Second-order Optimizer for Language Model Pre-training"
+(https://arxiv.org/abs/2305.14342) by Hong Liu, Zhiyuan Li, David Hall,
+Percy Liang, and Tengyu Ma.
+
+This contribution is heavily based on the implementation of Sophia by levanter
+(https://github.com/stanford-crfm/levanter) with some changes.
+"""
+from typing import Any, NamedTuple, Optional, Union, Callable
+
+import jax
+from jax import numpy as jnp
+from jax.random import PRNGKey
+
+from optax._src import base
+from optax._src import transform
+from optax._src.numerics import safe_int32_increment
+from optax._src.utils import canonicalize_dtype
+from optax._src.combine import chain
+from optax import tree_utils as otu
+
+
+class SophiaHState(NamedTuple):
+  """State for Sophia-H and similar."""
+
+  count: jax.Array  # shape=(), dtype=jnp.int32
+  mu: base.Updates  # momentum
+  nu: base.Updates  # EMA of hessian
+  key: PRNGKey
+
+
+def scale_by_sophia_h(
+    b1: float = 0.965,
+    b2: float = 0.99,
+    eps: float = 1e-8,
+    gamma: float = 0.01,
+    clip_threshold: Optional[float] = 1.0,
+    update_interval: int = 10,
+    mu_dtype: Optional[Any] = None,
+    pmap_axis_name: Optional[str] = None,
+    seed: PRNGKey = PRNGKey(0),
```

Comment:
I would probably make seed be either an int or a PRNGkey with default being seed=0.
If it is an int then convert it to a key later.

Reply:
Ok can do

Reply:
Oh this is another argument I changed to pass injecting hyperparams test, i think it has to be only one or the other, should it be a key or int?

Reply:
Let's make it an optional PNRGKey then. So if the user does not want to know all the details of the algorithm, he/she does not need to specify it (we may just instantiate it at the start if it is a None).